controller/volume_rebuilding_controller.go

```diff
@@ -192,12 +279,134 @@ func (vbc *VolumeRebuildingController) reconcile(volName string) (err error) {
 	}()
 
 	rebuildingAttachmentTicketID := longhorn.GetAttachmentTicketID(longhorn.AttacherTypeVolumeRebuildingController, volName)
+	if !vol.DeletionTimestamp.IsZero() {
+		vbc.logger.Infof("Volume %v is deleting, skip offline rebuilding", volName)
+		delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+		return nil
+	}
+	if vol.Status.Robustness == longhorn.VolumeRobustnessFaulted {
+		vbc.logger.Warnf("Volume %v is faulted, skip offline rebuilding", volName)
+		delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+		return nil
+	}
 
-	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+	isOfflineRebuildEnabled, err := vbc.isVolumeOfflineRebuildEnabled(vol.Spec.OfflineRebuilding)
+	if err != nil {
+		return err
+	}
+	if isOfflineRebuildEnabled && !util.IsHigherPriorityVATicketExisting(va, longhorn.AttacherTypeVolumeRebuildingController) {
+		if vol.Status.State == longhorn.VolumeStateDetached {
+			va, err = vbc.syncLHVolumeAttachmentForOfflineRebuild(vol, va, rebuildingAttachmentTicketID)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+		if vol.Status.State == longhorn.VolumeStateAttaching {
+			return nil
+		}
+
+		engine, err := vbc.getVolumeEngine(vol)
+		if err != nil {
+			return err
+		}
+		if engine == nil {
+			vbc.logger.Warnf("Volume %v engine not found, skip offline rebuilding", volName)
+			delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+			return nil
+		}
+		if engine.Status.ReplicaModeMap == nil {
+			// wait for engine status synced
+			return nil
+		}
+
+		isVolumeInRebuilding, err := vbc.isVolumeReplicasRebuilding(vol, engine)
+		if err != nil {
+			return err
```

Comment:
Do we need to remove the volumeattachment ticket if encountering an error?

Reply:
Added the defer function to delete the VA ticket if the following code flow returns an error.